jax/core.py

```diff
@@ -1216,6 +1239,30 @@ def canonicalize_shape(shape):
             "smaller subfunctions.")
   raise TypeError(msg.format(shape))
 
+def sorted_named_shape_tuples(named_shape_tuples):
+  return sorted(named_shape_tuples,
+                key=lambda x: x if isinstance(x, str) else hash(x))
+
+def canonicalize_named_shape(named_shape):
+  return dict(sorted_named_shape_tuples(named_shape.items()))
+
+def join_named_shapes(*named_shapes):
+  named_shape_tuples = sorted_named_shape_tuples(
+      item for named_shape in named_shapes for item in named_shape.items())
+  if not named_shape_tuples:
+    return {}
+  names, sizes = zip(*named_shape_tuples)
+  if len(set(names)) != len(names):
+    seen = set()
+    for name in names:
+      if name in seen:
+        break
+      seen.add(name)
+    raise TypeError(
+        f"Axis name {name} used with inconsistent sizes "
+        f"{[item[1] for item in named_shape_tuples if item[0] == name]}")
+  return dict(zip(names, sizes))
```

Comment:
I think if we called this function like

```python
join_named_shapes({'i': 3}, {'i': 3})
```

we'd get an error like

```
TypeError: Axis name i used with inconsistent sizes [3, 3]
```

The sizes are consistent so the error message is a little surprising.

I think the error check currently triggers any time the names on the inputs are not pairwise disjoint, regardless of sizes. Is that the intent, or rather do we want to allow inputs like the above example?

Reply:
Yup, I left out a `set()` call by mistake. Added a test.

Reply:
Btw I don't see the change yet in e5d9462, but let me know if I'm missing it.

Reply:
It went in 726ab16, I think.